Amqp.V1/Akka.Streams.Amqp.V1/AmqpSourceStage.cs

```diff
@@ -0,0 +1,68 @@
+﻿using Akka.Streams.Amqp.V1.Util;
+using Akka.Streams.Stage;
+using Akka.Streams.Supervision;
+using Amqp;
+using Amqp.Framing;
+using Amqp.Types;
+using System;
+
+namespace Akka.Streams.Amqp.V1
+{
+    public sealed class AmqpSourceStage<T> : GraphStage<SourceShape<T>>
+    {
+        public AmqpSourceStage(IAmqpSourceSettings<T> amqpSourceSettings)
+        {
+            AmqpSourceSettings = amqpSourceSettings;
+        }
+
+        public override SourceShape<T> Shape => new SourceShape<T>(Out);
+
+        public Outlet<T> Out { get; } = new Outlet<T>("AmqpSource.Out");
+        public IAmqpSourceSettings<T> AmqpSourceSettings { get; }
+
+        protected override GraphStageLogic CreateLogic(Attributes inheritedAttributes) => new AmqpSourceStageLogic(this, inheritedAttributes);
+
+        private class AmqpSourceStageLogic : GraphStageLogic
+        {
+            private readonly Outlet<T> outlet;
+            private readonly IAmqpSourceSettings<T> ampqSourceSettings;
+            private readonly ReceiverLink receiver;
+            private readonly Decider decider;
+
+            public AmqpSourceStageLogic(AmqpSourceStage<T> stage, Attributes attributes) : base(stage.Shape)
+            {
+                outlet = stage.Out;
+                ampqSourceSettings = stage.AmqpSourceSettings;
+                receiver = stage.AmqpSourceSettings.GetReceiverLink();
+                decider = attributes.GetDeciderOrDefault();
+
+                SetHandler(outlet, () => {
+                    var message = receiver.Receive();
+                    if (message != null)
+                    {
+                        try
+                        {
+                            Push(outlet, ampqSourceSettings.Convert(message));
+                            receiver.Accept(message);
```

Comment:
Is it a good idea to immediately accept the message ? 
What if I would want to accept it at the end of the stream because processing could fail later in the chain and if so I want to try to process it a second time?

Reply:
How would that work? Sending a reference to the receiverlink with the message.